gh-111178: Fix getsockaddrarg() undefined behavior

vstinner opened the pull request:
Don't pass direct references to sockaddr members since their type may not match PyArg_ParseTuple() types. Instead, use temporary 'int' and 'unsigned char' variables, and update sockaddr members afterwards.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-111178 -->
* Issue: gh-111178
<!-- /gh-issue-number -->

vstinner committed:
gh-111178: Fix getsockaddrarg() undefined behavior

Don't pass direct references to sockaddr members since their type may
not match PyArg_ParseTuple() types. Instead, use temporary 'int' and
'unsigned char' variables, and update sockaddr members afterwards.

picnixz commented on Modules/socketmodule.c:
```diff
@@ -2045,14 +2045,21 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             _BT_L2_MEMB(addr, bdaddr_type) = BDADDR_BREDR;
+            int psm;
+            int cid = _BT_L2_MEMB(addr, cid);
+            unsigned char bdaddr_type = _BT_L2_MEMB(addr, bdaddr_type);
```
Didn't we `bdaddr_type` just set to `BDADDR_BREDR` above?

vstinner replied:
> Didn't we bdaddr_type just set to BDADDR_BREDR above?

Oh, I didn't notice. I simplified the code (twice).

picnixz reviewed (commented):
Are `BTPROTO_RFCOMM` and `BTPROTO_HCI` for non-NetBSD below also suffering from the same issue? And yes, it is a bugfix IMO even if the function is internal.

vstinner committed:
Simplify code

vstinner committed:
Initialize cid to 0

vstinner committed:
Use 'unsigned short' type

Fix also BTPROTO_RFCOMM and BTPROTO_HCI code path.

vstinner committed:
Use 'unsigned char'

vstinner committed:
Add NEWS entry

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -2071,12 +2071,15 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             _BT_RC_MEMB(addr, family) = AF_BLUETOOTH;
-            if (!PyArg_ParseTuple(args, "si", &straddr,
-                                  &_BT_RC_MEMB(addr, channel))) {
+            unsigned char channel = _BT_RC_MEMB(addr, channel);
```
On Windows it is ULONG. See SOCKADDR_BTH_REDEF in Modules/socketmodule.h and https://learn.microsoft.com/en-us/windows/win32/api/ws2bth/ns-ws2bth-sockaddr_bth.

vstinner replied:
Alright, I modified the code to parse channel as 'unsigned long' on Windows.

vstinner committed:
Use 'unsigned long' to parse channel on Windows

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -2094,11 +2109,13 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
 #else  /* __NetBSD__ || __DragonFly__ */
             _BT_HCI_MEMB(addr, family) = AF_BLUETOOTH;
-            if (!PyArg_ParseTuple(args, "i", &_BT_HCI_MEMB(addr, dev))) {
+            unsigned short dev = _BT_HCI_MEMB(addr, dev);
```
Where is it defined?

picnixz replied:
There's a bunch of `define` at the top of the file:

```c
#if (defined(HAVE_BLUETOOTH_H) || defined(HAVE_BLUETOOTH_BLUETOOTH_H)) \
     && !defined(__NetBSD__) && !defined(__DragonFly__)
#define USE_BLUETOOTH 1
#if defined(__FreeBSD__)
...
#define _BT_HCI_MEMB(sa, memb) ((sa)->hci_##memb)
...
#elif defined(__NetBSD__) || defined(__DragonFly__) // <- unreachable
... 
#define _BT_HCI_MEMB(sa, memb) ((sa)->bt_##memb)
#else
...
#define _BT_HCI_MEMB(sa, memb) ((sa)->hci_##memb)
...
#endif
#endif
```

But AFAICT, the `elif defined(__NetBSD__) || defined(__DragonFly__)` is not possible at all since we're still in the big `#if` where we want `!defined(__NetBSD__) && !defined(__DragonFly__)`. So indeed, it looks like the macro wouldn't be defined here.

I think we should remove `!defined(__NetBSD__) && !defined(__DragonFly__)`

serhiy-storchaka replied:
I know about `_BT_HCI_MEMB`. I asked on what platform the `*_dev` member is defined, what its name and the structure name. Its turned out that it is `hci_node` on FreeBSD and NetBSD. `hci_dev` should exist on Linux.

vstinner committed:
Fix BTPROTO_HCI on FreeBSD

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -2107,6 +2107,21 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
+            }
+            const char *straddr = PyBytes_AS_STRING(args);
+            size_t len = PyBytes_GET_SIZE(args);
+            if (len >= sizeof(_BT_HCI_MEMB(addr, node))) {
```
I think that == is still valid, and you should use `strncpy`.

Test also that the bytes object does not contain embedded nulls.

vstinner replied:
> I think that == is still valid, and you should use strncpy.

Done. I wasn't sure since the manpage doesn't say anything about null character.

> Test also that the bytes object does not contain embedded nulls.

Ok, done.

serhiy-storchaka reviewed (commented):
You need also to make `makesockaddr` consisting with the parsing code.

vstinner committed:
Update BTPROTO_HCI code

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -2116,12 +2120,18 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             size_t len = PyBytes_GET_SIZE(args);
-            if (len >= sizeof(_BT_HCI_MEMB(addr, node))) {
+            if (strlen(straddr) != len) {
+                PyErr_Format(PyExc_OSError, "%s: "
```
Some code raises OSError for invalid address, other raises more traditional TypeError and ValueError.

I think that it is better to raise TypeError (for invalid type) and ValueError (for too long bytes or embedded null byte). They are programming or user data errors, not produced by the OS API call. In future we can change other OSError to more suitable exception types.

vstinner replied:
I changed the exception type to ValueError for "node too long" and "embedded null character".

serhiy-storchaka reviewed (commented):
LGTM, except that I am not sure about exception types.

vstinner committed:
Change exception type to ValueError

vstinner committed:
Remove outdated comments

vstinner committed:
Remove useless assignment

The socket address channel is not initialized before calling
PyArg_ParseTuple().
